crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,440 @@
+                CoinSigned,
+            },
+            contract::Contract,
+            message::{
+                MessageCoinPredicate,
+                MessageCoinSigned,
+                MessageDataPredicate,
+                MessageDataSigned,
+            },
+        },
+        ContractId,
+        Input,
+        Output,
+        TxId,
+        UtxoId,
+    },
+    services::txpool::PoolTransaction,
+};
+use petgraph::{
+    graph::NodeIndex,
+    prelude::StableDiGraph,
+    visit::EdgeRef,
+};
+
+use crate::{
+    collision_manager::CollisionReason,
+    error::Error,
+    ports::TxPoolDb,
+};
+
+use super::{
+    RemovedTransactions,
+    Storage,
+    StorageData,
+};
+
+pub struct GraphStorage {
+    /// The configuration of the graph
+    config: GraphConfig,
+    /// The graph of transactions
+    graph: StableDiGraph<StorageData, ()>,
+    /// Coins -> Transaction that crurrently create the UTXO
+    coins_creators: HashMap<UtxoId, NodeIndex>,
+    /// Contract -> Transaction that currently create the contract
+    contracts_creators: HashMap<ContractId, NodeIndex>,
+}
+
+pub struct GraphConfig {
+    /// The maximum number of transactions per dependency chain
+    pub max_txs_per_chain: u64,
+}
+
+impl GraphStorage {
+    /// Create a new graph storage
+    pub fn new(config: GraphConfig) -> Self {
+        Self {
+            config,
+            graph: StableDiGraph::new(),
+            coins_creators: HashMap::new(),
+            contracts_creators: HashMap::new(),
+        }
+    }
+}
+
+impl GraphStorage {
+    /// Remove a node and all its dependent sub-graph.
+    /// Edit the data of dependencies transactions accordingly.
+    /// Returns the removed transactions.
+    fn remove_node_and_dependent_sub_graph(
+        &mut self,
+        root: NodeIndex,
+    ) -> Result<Vec<PoolTransaction>, Error> {
+        let mut to_traverse = vec![root];
+        let mut to_remove = vec![];
+        while let Some(node_id) = to_traverse.pop() {
+            to_remove.push(node_id);
+            for edge in self
+                .graph
+                .edges_directed(node_id, petgraph::Direction::Outgoing)
+            {
+                to_traverse.push(edge.target());
+            }
+        }
+
+        let mut dependencies: Vec<NodeIndex> = self
+            .graph
+            .neighbors_directed(root, petgraph::Direction::Incoming)
+            .collect();
+        let nb_removed = to_remove.len() as u64;
+        let (gas_removed, tip_removed, removed_transactions) = to_remove.iter().fold(
+            (0, 0, vec![]),
+            |(gas, tip, mut removed_transactions), node_id| {
+                let Some(node) = self.graph.remove_node(*node_id) else {
+                    return (gas, tip, removed_transactions);
+                };
+                removed_transactions.push(node.transaction);
+                (
+                    gas + node.cumulative_gas,
+                    tip + node.cumulative_tip,
```

Comment:
This looks incorrect to me, if we sum the cumulative gas for a chain $A \rightarrow B \rightarrow C$ we'd end up with $\text{gas removed} = gas(A) + 2 gas(B) + 3 gas(C)$. I can't see any tests covering this, and changing this locally resulted in all tests passing for me.

Reply:
You are right. Thanks for catching this. It's updated.

I agree with you that the tests ported from the previous pool doesn't test if everything in the pool is well updated (caches, graph, ...).
This is something I want to work on in the next PR : https://github.com/FuelLabs/fuel-core/issues/2185
And I think it will showcase some bugs. Which is nice !

Reply:
Nice, yeah it makes sense not to do everything in this PR. It's huge as it is already :sweat_smile: